File: algorithms/0912.sort-an-array/solution.rb
# @param {Integer[]} nums
# @return {Integer[]}
def sort_array(nums)
  merge_sort(nums)
end

def merge_sort(arr)
  return arr if arr.size <= 1

  pivot = arr.size / 2

  left =  merge_sort(arr[0..pivot - 1])
  right = merge_sort(arr[pivot..-1])

  res = []

  while left.size > 0 && right.size > 0
    if left[0] < right[0]
      res << left.shift
    else
      res << right.shift
    end
  end

  while left.size > 0
    res << left.shift
  end

  while right.size > 0
    res << right.shift
  end

  res
end

File: algorithms/0912.sort-an-array/solution.cpp
#include <vector>

using namespace std;

class Solution {
  public:
    vector<int> sortArray(vector<int>& nums) { return mergeSort(nums); }

    vector<int> mergeSort(vector<int>& vec) {
        if (vec.size() <= 1) {
            return vec;
        }

        auto begin = vec.begin();
        auto end = vec.end();
        auto pivot = begin + vec.size() / 2;

        vector<int> v1(begin, pivot);
        vector<int> v2(pivot, end);

        vector<int> left = mergeSort(v1);
        vector<int> right = mergeSort(v2);

        vector<int> res;
        int l = 0;
        int r = 0;

        while (l < left.size() && r < right.size()) {
            if (left[l] < right[r]) {
                res.push_back(left[l]);
                l += 1;
            } else {
                res.push_back(right[r]);
                r += 1;
            }
        }

        while (l < left.size()) {
            res.push_back(left[l]);
            l += 1;
        }

        while (r < right.size()) {
            res.push_back(right[r]);
            r += 1;
        }

        return res;
    }
};
